File: Trinity.Runtime/platform_src/android/app/src/main/java/org/elastos/trinity/runtime/DIDSessionManager.java
package org.elastos.trinity.runtime;

import androidx.annotation.NonNull;

import org.json.JSONException;
import org.json.JSONObject;
import org.json.JSONStringer;

import java.io.Serializable;
import java.util.ArrayList;

public class DIDSessionManager {
    private WebViewActivity activity;
    private static DIDSessionManager instance;
    private AppManager appManager;

    static class IdentityAvatar {
        String contentType;
        byte[] base64ImageData;

        IdentityAvatar(String contentType, byte[] base64ImageData) {
            this.contentType = contentType;
            this.base64ImageData = base64ImageData;
        }

        public JSONObject asJsonObject() {
            try {
                JSONObject jsonObj = new JSONObject();
                jsonObj.put("contentType", contentType);
                jsonObj.put("base64ImageData", base64ImageData);
                return jsonObj;
            } catch (JSONException e) {
                e.printStackTrace();
                return null;
            }
        }

        public static IdentityAvatar fromJsonObject(JSONObject jsonObj) {
            if (!jsonObj.has("contentType") || !jsonObj.has("base64ImageData"))
                return null;

            try {
                IdentityAvatar avatar = new IdentityAvatar(
                        jsonObj.getString("contentType"),
                        (byte[]) jsonObj.get("base64ImageData"));

                return avatar;
            } catch (JSONException e) {
                e.printStackTrace();
                return null;
            }
        }
    }

    static class IdentityEntry {
        String didStoreId;
        String didString;
        String name;
        IdentityAvatar avatar;

        IdentityEntry(String didStoreId, String didString, String name) {
            this(didStoreId, didString, name, null);
        }

        IdentityEntry(String didStoreId, String didString, String name, IdentityAvatar avatar ) {
            this.didStoreId = didStoreId;
            this.didString = didString;
            this.name = name;
            this.avatar = avatar;
        }

        public JSONObject asJsonObject() {
            try {
                JSONObject jsonObj = new JSONObject();
                jsonObj.put("didStoreId", didStoreId);
                jsonObj.put("didString", didString);
                jsonObj.put("name", name);

                if (avatar != null) {
                    jsonObj.put("avatar", avatar.asJsonObject());
                }

                return jsonObj;
            } catch (JSONException e) {
                e.printStackTrace();
                return null;
            }
        }

        public static IdentityEntry fromJsonObject(JSONObject jsonObj) {
            if (!jsonObj.has("didStoreId") || !jsonObj.has("didString") || !jsonObj.has("name"))
                return null;

            try {
                IdentityEntry identity = new IdentityEntry(
                        jsonObj.getString("didStoreId"),
                        jsonObj.getString("didString"),
                        jsonObj.getString("name"));

                if (jsonObj.has("avatar")) {
                    identity.avatar = IdentityAvatar.fromJsonObject(jsonObj.getJSONObject("avatar"));
                }

                return identity;
            } catch (JSONException e) {
                e.printStackTrace();
                return null;
            }
        }
    }

    DIDSessionManager(WebViewActivity activity) {
        this.activity = activity;
        DIDSessionManager.instance = this;
    }

    public void setAppManager(AppManager appManager) {
        this.appManager = appManager;
    }

    public static DIDSessionManager getSharedInstance() {
        return instance;
    }

    public void addIdentityEntry(IdentityEntry entry) throws Exception {
        appManager.getDBAdapter().addDIDSessionIdentityEntry(entry);
    }

    public void deleteIdentityEntry(String didString) throws Exception {
        appManager.getDBAdapter().deleteDIDSessionIdentityEntry(didString);
    }

    public ArrayList<IdentityEntry> getIdentityEntries() throws Exception {
        return appManager.getDBAdapter().getDIDSessionIdentityEntries();
    }

    public IdentityEntry getSignedInIdentity() throws Exception {
        return appManager.getDBAdapter().getDIDSessionSignedInIdentity();
    }

    public void signIn(IdentityEntry identityToSignIn) throws Exception {
        // Make sure there is no signed in identity already
        DIDSessionManager.IdentityEntry signedInIdentity = DIDSessionManager.getSharedInstance().getSignedInIdentity();
        if (signedInIdentity != null)
            throw new Exception("Unable to sign in. Please first sign out from the currently signed in identity");

        appManager.getDBAdapter().setDIDSessionSignedInIdentity(identityToSignIn);

        // Ask the manager to handle the UI sign in flow.
        appManager.signIn();
    }

    public void signOut() throws Exception {
        appManager.getDBAdapter().setDIDSessionSignedInIdentity(null);

        // Ask the app manager to sign out and redirect user to the right screen
        appManager.signOut();
    }
}

File: Trinity.Runtime/platform_src/ios/elastOS/Classes/DIDSessions/DIDSessionManager.swift
/*
* Copyright (c) 2020 Elastos Foundation
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
*/

public class DIDSessionManager {
    private static var instance: DIDSessionManager? = nil
    private let dbAdapter: DIDSessionDatabaseAdapter
    private var appManager: AppManager? = nil

    init() {
        dbAdapter = DIDSessionDatabaseAdapter()
        DIDSessionManager.instance = self
    }

    public static func getSharedInstance() -> DIDSessionManager {
        return instance!
    }

    func setAppManager(_ appManager: AppManager) {
        self.appManager = appManager
    }
    
    func addIdentityEntry(entry: IdentityEntry) throws {
        _ = try dbAdapter.addDIDSessionIdentityEntry(entry: entry)
    }

    func deleteIdentityEntry(didString: String) throws {
        try dbAdapter.deleteDIDSessionIdentityEntry(didString: didString)
    }

    func getIdentityEntries() throws -> Array<IdentityEntry> {
        return try dbAdapter.getDIDSessionIdentityEntries()
    }

    func getSignedInIdentity() throws -> IdentityEntry? {
        return try dbAdapter.getDIDSessionSignedInIdentity()
    }

    func signIn(identityToSignIn: IdentityEntry) throws {
        // Make sure there is no signed in identity already
        guard (try DIDSessionManager.getSharedInstance().getSignedInIdentity()) == nil else {
            throw "Unable to sign in. Please first sign out from the currently signed in identity"
        }

        try dbAdapter.setDIDSessionSignedInIdentity(entry: identityToSignIn)

        // Ask the manager to handle the UI sign in flow.
        try appManager!.signIn()
    }

    public func signOut() throws {
        try dbAdapter.setDIDSessionSignedInIdentity(entry: nil)

        // Ask the app manager to sign out and redirect user to the right screen
        try appManager!.signOut()
    }
}
